File: update-html-from-csv.js
#!/usr/bin/env node

/**
 * Script to update index.html with data from act-list-template.csv
 * Run this script after editing the CSV file to update the HTML
 * Usage: node update-html-from-csv.js
 */

const fs = require('fs');
const path = require('path');

// Read CSV file
const csvPath = path.join(__dirname, 'act-list-template.csv');
const htmlPath = path.join(__dirname, 'index.html');

try {
    const csvContent = fs.readFileSync(csvPath, 'utf8');
    const lines = csvContent.split('\n').filter(line => line.trim());
    
    if (lines.length < 2) {
        console.log('CSV file is empty or has no data rows');
        process.exit(1);
    }
    
    // Detect delimiter
    const delimiter = lines[0].includes(';') ? ';' : ',';
    
    // Parse headers
    const headers = parseCSVLine(lines[0], delimiter).map(h => h.replace(/"/g, '').trim());
    
    // Parse data rows
    const data = [];
    for (let i = 1; i < lines.length; i++) {
        const values = parseCSVLine(lines[i], delimiter);
        if (values.length >= 5) {
            const item = {};
            headers.forEach((header, index) => {
                if (index < values.length) {
                    item[header] = values[index].replace(/^"|"$/g, '').trim();
                }
            });
            data.push(item);
        }
    }
    
    // Generate JavaScript data array
    const jsDataArray = data.map(item => {
        return `            { "Date": ${JSON.stringify(item['Date'] || '')}, "Location": ${JSON.stringify(item['Location'] || '')}, "Project Name Line 1": ${JSON.stringify(item['Project Name Line 1'] || '')}, "Project Name Line 2": ${JSON.stringify(item['Project Name Line 2'] || '')}, "Status": ${JSON.stringify(item['Status'] || '')}, "Clickable": ${JSON.stringify(item['Clickable'] || 'No')} }`;
    }).join(',\n');
    
    // Read HTML file
    let htmlContent = fs.readFileSync(htmlPath, 'utf8');
    
    // Find and replace the actListData array
    const dataPattern = /const actListData = \[[\s\S]*?\];/;
    const newData = `const actListData = [\n${jsDataArray}\n        ];`;
    
    if (dataPattern.test(htmlContent)) {
        htmlContent = htmlContent.replace(dataPattern, newData);
        fs.writeFileSync(htmlPath, htmlContent, 'utf8');
        console.log(`✅ Successfully updated HTML with ${data.length} act items from CSV`);
    } else {
        console.log('❌ Could not find actListData in HTML file');
        process.exit(1);
    }
    
} catch (error) {
    console.error('❌ Error:', error.message);
    process.exit(1);
}

// CSV line parser
function parseCSVLine(line, delimiter = ',') {
    const result = [];
    let current = '';
    let inQuotes = false;
    
    for (let i = 0; i < line.length; i++) {
        const char = line[i];
        if (char === '"') {
            inQuotes = !inQuotes;
        } else if (char === delimiter && !inQuotes) {
            result.push(current);
            current = '';
        } else {
            current += char;
        }
    }
    result.push(current);
    return result;
}

File: update-html-from-csv.py
#!/usr/bin/env python3
"""
Script to update index.html with data from act-list-template.csv
Run this script after editing the CSV file to update the HTML
Usage: python3 update-html-from-csv.py
"""

import csv
import re
import json
import os

# File paths
csv_path = 'act-list-template.csv'
html_path = 'index.html'

try:
    # Read CSV file
    with open(csv_path, 'r', encoding='utf-8') as f:
        # Detect delimiter
        first_line = f.readline()
        delimiter = ';' if ';' in first_line else ','
        f.seek(0)
        
        # Read CSV
        reader = csv.DictReader(f, delimiter=delimiter)
        data = list(reader)
    
    if not data:
        print('CSV file is empty or has no data rows')
        exit(1)
    
    # Generate JavaScript data array
    js_items = []
    for item in data:
        js_item = (
            f'            {{ "Date": {json.dumps(item.get("Date", ""))}, '
            f'"Code/ID": {json.dumps(item.get("Code/ID", ""))}, '
            f'"Description Line 1": {json.dumps(item.get("Description Line 1", ""))}, '
            f'"Description Line 2": {json.dumps(item.get("Description Line 2", ""))}, '
            f'"Status": {json.dumps(item.get("Status", ""))}, '
            f'"Clickable": {json.dumps(item.get("Clickable", "No"))} }}'
        )
        js_items.append(js_item)
    
    js_data_array = ',\n'.join(js_items)
    
    # Read HTML file
    with open(html_path, 'r', encoding='utf-8') as f:
        html_content = f.read()
    
    # Find and replace the actListData array
    pattern = r'const actListData = \[[\s\S]*?\];'
    new_data = f'const actListData = [\n{js_data_array}\n        ];'
    
    if re.search(pattern, html_content):
        html_content = re.sub(pattern, new_data, html_content)
        with open(html_path, 'w', encoding='utf-8') as f:
            f.write(html_content)
        print(f'✅ Successfully updated HTML with {len(data)} act items from CSV')
    else:
        print('❌ Could not find actListData in HTML file')
        exit(1)
        
except FileNotFoundError as e:
    print(f'❌ File not found: {e.filename}')
    exit(1)
except Exception as e:
    print(f'❌ Error: {e}')
    exit(1)

